### PR TITLE
Convert common time helpers to Typescript

### PR DESCRIPTION
**Pull request by lhsazevedo**
See flarum/framework#3533

### CODE REVIEW

File: js/src/common/helpers/fullTime.tsx [hunk 1]
```diff
@@ -1,11 +1,11 @@
+import dayjs from 'dayjs';
+import * as Mithrill from 'mithril';
```
Comment (askvortsov1) [inline js/src/common/helpers/fullTime.tsx:2]: Typo

**Review by askvortsov1 — CHANGES_REQUESTED** [review]
This mostly makes sense, thanks!

**Review by clarkwinkelmann — COMMENTED** [review]
Why does one import uses `import * as dayjs from 'dayjs'; ` while the others use `import dayjs from 'dayjs'; ` ?

Do we actually want to import global variables or could we just type-hint them somewhere? Using the global variable instead of an import will save up on the file size, as webpack won't have to add the additional code that resolves each import.

**Comment by datitisev** [comment]
@clarkwinkelmann We can test locally if the bundle size increases if they're just used for typings. Haven't done so myself yet.

File: js/src/common/helpers/fullTime.tsx (hunk 1 above)
Comment (lhsazevedo) [reply js/src/common/helpers/fullTime.tsx]: Fixed

**Comment by lhsazevedo** [comment]
I removed the `import * as dayjs` variant and now all three changed files use the default export

File: js/src/common/helpers/fullTime.tsx (hunk 1 above)
Comment (askvortsov1) [reply js/src/common/helpers/fullTime.tsx]: Thanks! :D

**Review by askvortsov1 — APPROVED** [review]
Looks good to me! I'm approving this as long as

> Do we actually want to import global variables or could we just type-hint them somewhere? Using the global variable instead of an import will save up on the file size, as webpack won't have to add the additional code that resolves each import.

doesn't prove to be an issue

**Comment by lhsazevedo** [comment]
@askvortsov Thanks, I'll test this as soon as possible.

**Comment by lhsazevedo** [comment]
## Test
So, apparently webpack doesn't inject the module resolving code on each module, but use a global one. [Here's the test i did](https://gist.github.com/lhsazevedo/a49c5962db67251313e0b22ebbed4d37).

## Flarum build
Heres `src/common/utils/humanTime.ts` unminified output:
```js
"use strict";
__webpack_require__.d(__webpack_exports__, "a", function() { return humanTime; });
var dayjs = __webpack_require__(28);
var dayjs_default = /*#__PURE__*/__webpack_require__.n(dayjs);
var dayjs_plugin_relativeTime = __webpack_require__(73);
var dayjs_plugin_relativeTime_default = /*#__PURE__*/__webpack_require__.n(dayjs_plugin_relativeTime);

function humanTime(time) {
  var d = dayjs_default()(time);
  var now = dayjs_default()();

  // ...
}
```

## Dayjs plugins

One thing we could optimize is, even not using directly, we need to import relativeTime dayjs plugin at `js/src/common/utils/humanTime.ts`. Without this import, typescript can't pick up tipes for the [fromNow()](https://github.com/flarum/core/blob/2ff46f5a2ebe2532f780f6e60a921beb41876d69/js/src/common/utils/humanTime.ts#L32) method and complains.
The code still works though, because relativeTime plugin is installed at `js/src/common/index.js`:

https://github.com/flarum/core/blob/2ff46f5a2ebe2532f780f6e60a921beb41876d69/js/src/common/index.js#L12-L16

We can avoid this import if we move this plugin installation process to a separated dayjs library module, something like:
```typescript
// js/src/common/lib/dayjs.ts

import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime';

dayjs.extend(relativeTime)

export default dayjs


// js/src/common/utils/humanTime.ts

// no need import relativeTime from 'dayjs/plugin/relativeTime';
import dayjs from '../../common/lib/dayjs'
```
~Should I open a new issue for this?~ See flarum/framework#2417

**Comment by askvortsov1** [comment]
> So, apparently webpack doesn't inject the module resolving code on each module, but use a global one. [Here's the test i did](https://gist.github.com/lhsazevedo/a49c5962db67251313e0b22ebbed4d37).

Thank you for testing this!



> We can avoid this import if we move this plugin installation process to a separated dayjs library module, something like:

Hmm I really like this idea! @clarkwinkelmann @rob006 could this also work for https://discuss.flarum.org/d/25196-beta-14-dayjs-translations?

**Comment by datitisev** [comment]
@askvortsov1 It could technically work but the config files would need to use `flarum.core.compat['flarum/lib/dayjs']` to access it.

**Comment by lhsazevedo** [comment]
It seems that it would be a more drastic change, and the code already works without it, so I created [a new issue](https://github.com/flarum/core/issues/2417) for this subject so that we can move forward here.

**Review by clarkwinkelmann — APPROVED** [review]
I'm triggered by the imports order ahah (`fullTime` imports dayjs first, while `humanTime` imports Mithril first :hear_no_evil: )

Looks good otherwise.

**Comment by lhsazevedo** [comment]
@clarkwinkelmann Fixed following this rule: external imports first, internal last, both alphabetically sorted.

**Review by tankerkiller125 — APPROVED** [review]

**Pull request merged**